docs/source/format/CDataInterface.rst

```diff
@@ -0,0 +1,752 @@
+   named ``entry``, itself a 2-child struct type of ``(key, value)``.
+
+Examples
+--------
+
+* A dictionary-encoded ``decimal128(precision = 12, scale = 5)`` array
+  with ``int16`` indices has format string ``s``, and its dependent dictionary
+  array has format string ``d:12,5``.
+* A ``list<uint64>`` array has format string ``+l``, and its single child
+  has format string ``L``.
+* A ``struct<ints: int32, floats: float32>`` has format string ``+s``; its two
+  children have names ``ints`` and ``floats``, and format strings ``i`` and
+  ``f`` respectively.
+* A ``map<string, float64>`` array has format string ``+m``; its single child
+  has name ``entry`` and format string ``+s``; its two grandchildren have names
+  ``key`` and ``value``, and format strings ``u`` and ``g`` respectively.
+* A ``sparse_union<ints: int32, floats: float32>`` with type ids ``4, 5``
+  has format string ``+us:4,5``; its two children have names ``ints`` and
+  ``floats``, and format strings ``i`` and ``f`` respectively.
+
+
+Structure definitions
+=====================
+
+The following free-standing definitions are enough to support the Arrow
+C data interface in your project.  Like the rest of the Arrow project, they
+are available under the Apache License 2.0.
+
+.. code-block:: c
+
+   #define ARROW_FLAG_ORDERED 1
+   #define ARROW_FLAG_NULLABLE 2
+
+   struct ArrowArray {
+     // Type description
+     const char* format;
+     const char* name;
+     const char* metadata;
+     int64_t flags;
+
+     // Data description
+     int64_t length;
+     int64_t null_count;
+     int64_t offset;
+     int64_t n_buffers;
+     int64_t n_children;
+     const void** buffers;
+     struct ArrowArray** children;
+     struct ArrowArray* dictionary;
+
+     // Release callback
+     void (*release)(struct ArrowArray*);
+     // Opaque producer-specific data
+     void* private_data;
+   };
+
+The ArrowArray structure
+------------------------
+
+.. c:member:: const char* ArrowArray.format
+
+   Mandatory.  A null-terminated, UTF8-encoded string describing
+   the data type.  If the data type is nested, child types are not
+   encoded here but in the :c:member:`ArrowArray.children` arrays.
+
+   Consumers MAY decide not to support all data types, but they
+   should document this limitation.
+
+.. c:member:: const char* ArrowArray.name
+
+   Optional.  A null-terminated, UTF8-encoded string of the field
+   or array name.  This is mainly used to reconstruct child fields
+   of nested arrays.
+
+   Producers MAY decide not to provide this information, and consumers
+   MAY decide to ignore it.  If omitted, MAY be NULL or an empty string.
+
+.. c:member:: const char* ArrowArray.metadata
+
+   Optional.  A null-terminated, UTF8-encoded string describing
+   the type's metadata.  If the data type is nested, child types are not
+   encoded here but in the :c:member:`ArrowArray.children` arrays.
+
+   MUST be a JSON-compatible mapping of UTF8 strings to UTF8 strings.
```

Comment:
Did any else express opinions on JSON?  It still feels strange to me to require a JSON parser given everything the structure of the rest of metadata.

Reply:
I would be ok with something else but I can't think of something simple enough. Do you have something in mind? Perhaps a large concatenated sequence of "(key, value)" null-terminated strings?

(e.g. `one\0two\0three\0four\0'` would represent metadata `{'one': 'two', 'three': 'four'}`)

Reply:
> Do you have something in mind? Perhaps a large concatenated sequence of "(key, value)" null-terminated strings?

This to strings with length prepended as int32_t (the one advantage of having lengths is being able to skip more quickly over values that aren't relevant.

Reply:
Could do a simple encoding like

```
int32: number of key/value pairs
[KeyValue]

where KeyValue is

int32: key size
uint8_t*: key
int32: value size
uint8_t*: value
```

Has to be parsed (and need to stipulate little-endianness) but it's simpler than JSON

Reply:
+1 for @wesm's proposal.

Reply:
I'll give it a try in the draft implementation (PR #5608).

Reply:
Ok, I implemented it. Turns out to be quite easy. I'll change the spec soon.